### PR TITLE
Feature: Max Items with current Purse

### PR DESCRIPTION
**Pull request by NetheriteMiner**
Depend on
- #871

When viewing an item in the bazaar, draws an overlay that shows the maximum number of items you could buy with your current purse for buy orders and buying instantly.

Currently, if your purse is changing because you just got some coins (from a talisman, most likely), the GUI will briefly not render because the regex that I use for determining the number of coins in your purse was defined in DiscordStatus, and hasn't been updated to use the repo nor does it account for the (+some number) at the end if your purse is changing. I intend to make another PR fixing that soon.

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt]: Do not read the purse from the scoreboard, use the PurseAPI instead.

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
looking good apart from formatting of the loops and matching the patterns.
if you wanted to you could also make the patterns be part of a group and formatted like https://github.com/hannibal002/SkyHanni/pull/871 but that is optional just makes it more readable

File: src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt]: matchMatcher already ensures matcher.matches is true so this can reduce stuff like lore@ and items@ in your code so that it is less indented and more readable

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
forgot to send the comment before my previous review :(

File: src/main/java/at/hannibal2/skyhanni/config/features/inventory/InventoryConfig.java
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/inventory/InventoryConfig.java]:
```suggestion
    @Expose
    public Position purseItemsPos = new Position(30, 10, true, false);
    
    @Expose
```

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
another small thing to change

**Comment by hannibal002** [comment]
Let's wait until #871 is merged for this.

File: src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt]: you can use `group("coins").formatNumber()` instead

File: src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt]: use long instead of float: `private var buyOrderPrice: Long? = null`

File: src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt:46]: we can check if item.displayName is null at this spot already, continuing even earlier.

File: src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt]: use nullable instead of -1 to check for default value

File: src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt]: Maybe use `val buyInstant = (currentPurse / instantBuyPrice).toInt().coerceAtLeast(0)` instead for those two cases

File: src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt]: Is there a reason for casting to int?

File: src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt]: Please move the two usages of `SkyHanniMod.feature.inventory` to a config member for consistency.

**Review by hannibal002 — CHANGES_REQUESTED** [review]
I think the feature display can use a short description title so it's clear that this does not describe the price but rather a number.
Please give the display color. also, maybe change the default position.
image

File: src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt
Comment (Rohan-Sangani) [reply src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt]: Unfortunately, I can't because String.formatNumber returns a Long, I need the prices to be floats or doubles since losing the fractional portion will mess up the calculations - same applies to your other suggestions

File: src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt]: please have all of these plit across 4 lines like the `instantPattern` already does

File: src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt]: these should be gone in final version

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
two small things i saw

File: src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt]:
```suggestion
    private val patternGroup by RepoPattern.group("inventory.maxpurse")
    private val orderPattern by patternGroup.pattern(
        "order",
        ".*§6(?<coins>[\\d.,]+) coins §7each.*"
    )
    private val instantPattern by patternGroup.pattern(
        "instant",
        ".*Price per unit: §6(?<coins>[\\d.,]+) coins.*"
    )
    private val createOrderPattern by patternGroup.pattern(
        "createorder",
        "§aCreate Buy Order"
    )
    private val createInstantPattern by patternGroup.pattern(
        "createinstant",
        "§aBuy Instantly"
    )
    
```

I did make this change on the website so this might not 100% work so best to test it as well but should be using a pattern group

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
I also missed seeing this

File: src/main/java/at/hannibal2/skyhanni/data/PurseAPI.kt
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/data/PurseAPI.kt]: Please do not remove the empty space here

File: src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt]: please use `group("coins").formatDouble() + 0.1` instead

File: src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/MaxPurseItems.kt]: here as well

**Review by hannibal002 — CHANGES_REQUESTED** [review]
I don't like this color/formatting. It does not feel like SkyHanni to me. Please use gray or yellow or another color more prominently used in SkyHanni for the text, and yellow for the number. also, you can split it up into three or more lines:

"with your current purse, you can buy
x times via buy order
or x times via instant buy
from bazaar"

Or something similar.
Just some ideas.

grafik

**Review by CalMWolfs — APPROVED** [review]
Changes look good, haven't tested in game

**Review by hannibal002 — APPROVED** [review]

**Pull request merged**
